components/message.php

```diff
@@ -0,0 +1,504 @@
+		'sender_id',
+		'date_sent',
+	);
+
+	/**
+	 * Add a message.
+	 *
+	 * ## OPTIONS
+	 *
+	 * --from=<user>
+	 * : Identifier for the user. Accepts either a user_login or a numeric ID.
+	 *
+	 * [--to=<user>]
+	 * : Identifier for the recipient. Accepts either a user_login or a numeric ID.
+	 * ---
+	 * Default: Empty.
+	 * ---
+	 *
+	 * [--subject=<subject>]
+	 * : Subject of the message.
+	 * ---
+	 * Default: Message Subject.
+	 * ---
+	 *
+	 * [--content=<content>]
+	 * : Content of the message.
+	 * ---
+	 * Default: Random content.
+	 * ---
+	 *
+	 * [--thread-id=<thread-id>]
+	 * : Thread ID.
+	 * ---
+	 * Default: false
+	 * ---
+	 *
+	 * [--date-sent=<date-sent>]
+	 * : MySQL-formatted date.
+	 * ---
+	 * Default: current date.
+	 * ---
+	 *
+	 * [--silent=<silent>]
+	 * : Whether to silent the message creation.
+	 * ---
+	 * Default: false.
+	 * ---
+	 *
+	 * [--porcelain]
+	 * : Return the thread id of the message.
+	 *
+	 * ## EXAMPLES
+	 *
+	 *     $ wp bp message create --from=user1 --to=user2 --subject="Message Title" --content="We are ready"
+	 *     Success: Message successfully created.
+	 *
+	 *     $ wp bp message create --from=545 --to=313
+	 *     Success: Message successfully created.
+	 *
+	 * @alias add
+	 */
+	public function create( $args, $assoc_args ) {
+		$r = wp_parse_args( $assoc_args, array(
+			'to'        => '',
+			'subject'   => '',
+			'content'   => '',
+			'thread-id' => false,
+			'date-sent' => bp_core_current_time(),
+			'silent'    => false,
+		) );
+
+		// Fill in the content.
+		if ( empty( $r['content'] ) ) {
+			$r['content'] = $this->generate_random_text();
+		}
+
+		// Fill in the subject.
+		if ( empty( $r['subject'] ) ) {
+			$r['subject'] = sprintf( 'Message Subject' );
+		}
+
+		$user = $this->get_user_id_from_identifier( $assoc_args['from'] );
+		if ( ! $user ) {
+			WP_CLI::error( 'No user found by that username or ID.' );
+		}
+
+		// To is not required when thread id is set.
+		if ( ! empty( $r['to'] ) ) {
+			$recipient = $this->get_user_id_from_identifier( (int) $r['to'] );
+
+			if ( ! $recipient ) {
+				WP_CLI::error( 'No user found by that username or ID.' );
+			}
+		}
+
+		// Existing thread recipients will be assumed.
+		$recipient = ( ! empty( $r['thread_id'] ) ) ? array() : array( $recipient->ID );
+
+		$thread_id = messages_new_message( array(
+			'sender_id'  => $user->ID,
+			'thread_id'  => $r['thread-id'],
+			'recipients' => $recipient,
+			'subject'    => $r['subject'],
+			'content'    => $r['content'],
+			'date_sent'  => $r['date-sent'],
+		) );
+
+		if ( ! is_numeric( $thread_id ) ) {
+			WP_CLI::error( 'Could not add a message.' );
+		}
+
+		if ( $r['silent'] ) {
+			return;
+		}
+
+		if ( \WP_CLI\Utils\get_flag_value( $assoc_args, 'porcelain' ) ) {
+			WP_CLI::line( $thread_id );
+		} else {
+			WP_CLI::success( 'Message successfully created.' );
+		}
+	}
+
+	/**
+	 * Delete message thread(s) for a given user.
+	 *
+	 * ## OPTIONS
+	 *
+	 * <thread-id>...
```

Comment:
The semantics are confusing here. `wp bp message delete 123` only works if `123` is a *thread* ID, not a member ID. Perhaps the subcommand should be `delete-thread`.

A `delete` command for individual messages would be nice, but it looks like BP doesn't actually have this (?!?).